src/server/mod.rs

```diff
@@ -966,14 +980,274 @@ impl LanguageServer for LspServer {
+                    match parent {
+                        Some(AstNode::ImportDeclaration(_)) => {
+                            let infos: Vec<(String, String)> =
+                                if let Some(env) = flux::imports() {
+                                    env.iter().filter(|(path, _val)| {
+                                    crate::shared::get_package_name(path).is_some()
+                                }).map(|(path, _val)| {
+                                    #[allow(clippy::expect_used)]
+                                    (crate::shared::get_package_name(path).expect("Previous filter failed.").into(), path.clone())
+                                }).collect()
+                                } else {
+                                    return Ok(None);
+                                };
+                            let imports =
+                                completion::get_imports(&sem_pkg);
+
+                            infos.into_iter().filter(|(name, _path)| {
+                                !&imports.iter().any(|x| &x.path == name)
+                            }).map(|(_name, path)| {
+                                let trigger = if let Some(context) = & params.context {
+                                    context.trigger_character.as_deref()
+                                } else {
+                                    None
+                                };
+                                let insert_text = if trigger == Some("\"") {
+                                    path.as_str().to_string()
+                                } else {
+                                    format!(r#""{}""#, path.as_str())
+                                };
+                                lsp::CompletionItem {
+                                    label: insert_text.clone(),
+                                    insert_text: Some(insert_text),
+                                    insert_text_format: Some(lsp::InsertTextFormat::SNIPPET),
+                                    kind: Some(lsp::CompletionItemKind::VALUE),
+                                    ..lsp::CompletionItem::default()
+                                }
+                            }).collect()
+                        }
+                        // This is where bucket/measurement/field/tag completion will occur.
+                        Some(_) | None => return Ok(None),
+                    }
+                }
+                _ => return Ok(None),
+            },
+            None => return Ok(None),
+        };
+        if items.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(lsp::CompletionResponse::List(
+                lsp::CompletionList {
+                    // XXX: rockstar (5 Jul 2022) - This should probably always be incomplete, so
+                    // we don't leave off to the client to try and figure out what completions to use.
+                    is_incomplete: false,
```

Comment:
Yeah, we have a class of bugs that seem related to expecting the client to be smarter about how it presents completions. The hypothesis currently is that when we send `is_incomplete: false` for a specific context, the client (specifically the monaco client, in vsflux and the browser) caches that data and doesn't ask the server again. Unfortunately, it doesn't know flux, and its guesses are sometimes bad. It _could_ be that we're setting some flags incorrectly in our completions as well, so I'm not willing to go down that path completely yet, but it's the working theory.

I added this comment specifically because I was going to knee-jerk change it, but caught myself because I shouldn't be changing behavior in this refactor patch. Adding the comment made me feel better. 😄